File: spring-graphql-annotations/src/main/java/com/example/demo/service/PostService.java
package com.example.demo.service;

import com.example.demo.gql.types.*;
import com.example.demo.model.CommentEntity;
import com.example.demo.model.PostEntity;
import com.example.demo.repository.AuthorRepository;
import com.example.demo.repository.CommentRepository;
import com.example.demo.repository.PostRepository;
import lombok.RequiredArgsConstructor;
import org.apache.commons.lang3.Validate;
import org.reactivestreams.Publisher;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Sinks;

import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.function.Function;

@Service
@RequiredArgsConstructor
public class PostService {
    private static final Function<PostEntity, Post> POST_MAPPER = p -> Post.builder()
            .id(p.id().toString())
            .title(p.title())
            .content(p.content())
            .status(PostStatus.valueOf(p.status()))
            .createdAt(p.createdAt())
            .authorId(p.authorId().toString())
            .build();
    public static final Function<CommentEntity, Comment> COMMENT_MAPPER = c -> Comment.builder()
            .id(c.id().toString())
            .content(c.content())
            .createdAt(c.createdAt())
            .postId(c.postId().toString())
            .build();
    final PostRepository posts;
    final CommentRepository comments;
    final AuthorRepository authors;


    public List<Post> getAllPosts() {
        return this.posts.findAll()
                .stream()
                .map(POST_MAPPER)
                .toList();
    }

    public Post getPostById(String id) {
        var postEntity = this.posts.findById(UUID.fromString(id));
        return Optional.ofNullable(postEntity)
                .map(POST_MAPPER)
                .orElseThrow(() -> new PostNotFoundException(id));
    }

    List<Post> getPostsByAuthorId(String id) {
        return this.posts.findByAuthorId(UUID.fromString(id))
                .stream()
                .map(POST_MAPPER)
                .toList();
    }

    public Post createPost(CreatePostInput postInput) {
        Validate.notNull(postInput, "CreatePostInput can not be null");
        Validate.notEmpty(postInput.getTitle(), "CreatePostInput.title can not be empty");
        // Use a hard code user id here.
        // In a real world application, the author is the current user which can be fetched from Spring security context.
        var authorId = this.authors.findAll().get(0).id();
        UUID id = this.posts.create(postInput.getTitle(), postInput.getContent(), "DRAFT", authorId);
        var entity = this.posts.findById(id);
        return Optional.ofNullable(entity)
                .map(POST_MAPPER)
                .orElseThrow(() -> new PostNotFoundException(id.toString()));
    }

    public List<Comment> getCommentsByPostId(String id) {
        return this.comments.findByPostId(UUID.fromString(id))
                .stream()
                .map(COMMENT_MAPPER)
                .toList();
    }

    public List<Comment> getCommentsByPostIdIn(Set<String> ids) {
        var uuids = ids.stream().map(UUID::fromString).toList();
        return this.comments.findByPostIdIn(uuids)
                .stream()
                .map(COMMENT_MAPPER)
                .toList();
    }

    public Comment addComment(CommentInput input) {
        UUID id = this.comments.create(input.getContent(), UUID.fromString(input.getPostId()));
        Comment commentById = this.getCommentById(id.toString());
        sink.emitNext(commentById, Sinks.EmitFailureHandler.FAIL_FAST);
        return commentById;
    }

    // sink of `commentAdded` event
    private Sinks.Many<Comment> sink = Sinks.many().replay().latest();

    // subscription for `commentAdded`
    public Publisher<Comment> commentAdded() {
        return sink.asFlux();
    }

    public Comment getCommentById(String id) {
        var commentById = this.comments.findById(UUID.fromString(id));
        return COMMENT_MAPPER.apply(commentById);
    }

}

File: dgs-kotlin/src/main/kotlin/com/example/demo/service/PostService.kt
package com.example.demo.service

import com.example.demo.asGqlType
import com.example.demo.gql.types.Comment
import com.example.demo.gql.types.CommentInput
import com.example.demo.gql.types.CreatePostInput
import com.example.demo.gql.types.Post
import com.example.demo.model.CommentEntity
import com.example.demo.model.PostEntity
import com.example.demo.repository.CommentRepository
import com.example.demo.repository.PostRepository
import org.reactivestreams.Publisher
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import reactor.core.publisher.Sinks
import java.util.*

@Service
class PostService(
    val posts: PostRepository,
    val comments: CommentRepository
) {
    companion object{
        private val log = LoggerFactory.getLogger(PostService::class.java)
    }

    fun allPosts() = this.posts.findAll().map { it.asGqlType() }

    fun getPostById(id: UUID): Post = this.posts.findById(id)
        .map { it.asGqlType() }
        .orElseThrow { PostNotFoundException(id) }

    fun getPostsByAuthorId(id: UUID) = this.posts.findByAuthorId(id).map { it.asGqlType() }

    fun createPost(postInput: CreatePostInput): Post {
        val data = PostEntity(title = postInput.title, content = postInput.content)
        val saved = this.posts.save(data)
        return saved.asGqlType()
    }

    fun addComment(commentInput: CommentInput): Comment {
        val postId = commentInput.postId
        return this.posts.findById(postId)
            .map {
                val data = CommentEntity(content = commentInput.content, postId = postId)
                val saved = this.comments.save(data)
                val comment = saved.asGqlType()
                log.debug("emitting $comment to event `commentAdded`")
                sink.emitNext(comment, Sinks.EmitFailureHandler.FAIL_FAST)
                comment
            }
            .orElseThrow { PostNotFoundException(postId) }
    }

    val sink = Sinks.many().replay().latest<Comment>()
    fun commentAdded(): Publisher<Comment> = sink.asFlux()

    fun getCommentsByPostId(id: UUID): List<Comment> = this.comments.findByPostId(id)
        .map { it.asGqlType() }

    fun getCommentsByPostIdIn(ids: List<UUID>): List<Comment> {
        val authorEntities = comments.findByPostIdIn(ids)
        return authorEntities.map { it.asGqlType() }
    }
}